datalad/support/gitrepo.py

```diff
@@ -3444,6 +3468,49 @@ def save_(self, message=None, paths=None, _status=None, **kwargs):
         # TODO yield result for commit, prev helper checked hexsha pre
         # and post...
 
+    def _check_for_win_compat(self, files):
+        """Check file names for illegal characters or reserved names on Windows
+
+        In the case that a non-Windows-compatible file is detected, warn users
+        about potential interoperability issues.
+        """
+        from collections import defaultdict
+        problems = defaultdict(list)
+        for file in files.keys():
+            for part in Path(file).parts:
```

Comment:
I worry a bit about performance overhead from this loop on heavy in number of files datasets.  I wonder if any timing was done?

Reply:
No I haven't. What number of files should I test? Do you by any chance have an example benchmarking procedure with an increasing number of files?

Reply:
we do not have anything formal unfortunately at the needed scale within our benchmarks :-/  I guess it would have been valuable to add a benchmark to "save 1000 files" or alike -- could come handy in any other use case when we change `save` functionality.

Reply:
Also no formal comparison, but I ran the following script to general 10000 files and one incompatible file and save everything a few times:

```
#! /bin/bash
DIR=$(mktemp -d)
cd $DIR && datalad create -f .
echo "Writing 10000 files..."
for n in {1..10000}; do
    dd if=/dev/urandom of=file$( printf %03d "$n" ).bin bs=1 count=$(( RANDOM + 1024 )) status=none
done
echo "incompatible" > COM1.txt
time datalad save .
```

Outside of this branch it yields:

```
[INFO   ] Creating a new annex repo at /tmp/tmp.zf6J7c3bg0 
create(ok): /tmp/tmp.zf6J7c3bg0 (dataset)
Writing 10000 files...
add(ok): COM1.txt (file)                                                        
add(ok): file001.bin (file)                                                     
add(ok): file002.bin (file)                                                     
add(ok): file003.bin (file)                                                     
add(ok): file004.bin (file)                                                     
add(ok): file005.bin (file)                                                     
add(ok): file006.bin (file)                                                     
add(ok): file007.bin (file)                                                     
add(ok): file008.bin (file)                                                     
add(ok): file009.bin (file)                                                     
  [1 similar message has been suppressed; disable with datalad.ui.suppress-simil  [9991 similar messages have been suppressed; disable with datalad.ui.suppress-similar-results=off]
save(ok): . (dataset)                                                           
action summary:                                                                 
  add (ok: 10001)
  save (ok: 1)

real	0m23.287s
user	0m18.608s
sys	0m9.586s
bash wincompat  139.60s user 555.50s system 100% cpu 11:31.80 total
```

With the config set to ``warning`` I saw
```
action summary:                                                                                                   
  add (ok: 10001)
  save (ok: 1)

real	0m22.583s
user	0m18.283s
sys	0m9.228s
DATALAD_SAVE_WINDOWS__COMPAT__WARNING='warning' bash wincompat  147.58s user 593.18s system 95% cpu 12:56.23 total
```

with ``error``

```

real	0m23.890s
user	0m19.235s
sys	0m10.133s
DATALAD_SAVE_WINDOWS__COMPAT__WARNING='error' bash wincompat  148.41s user 895.40s system 100% cpu 17:20.15 total
```

and with ``none``

```

real	0m29.357s
user	0m23.892s
sys	0m12.222s
DATALAD_SAVE_WINDOWS__COMPAT__WARNING='none' bash wincompat  154.34s user 907.61s system 100% cpu 17:37.29 total
```

(a bit strange that the time increases without the check, but I hope that's just chance - I think the important benchmark is that there is no increase in time between the first and 2nd/3rd timings?)

Reply:
those trailing `... bash wincompat  ... 11:31.80 total` to be ignored? (they seems to be high -- minutes, right?)

Reply:
I just wrote a (probably pretty inefficient) script that takes so long

Reply:
I will mark this conversation resolved.  We have optimized it a bit, so I should sleep better ;)